src/mon/AuthMonitor.cc

```diff
@@ -1630,34 +1630,47 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
 	}
       }
 
+      mon_cap_string += mon_cap_string.empty() ? "" : ", ";
+      string mon_perm = cap;
+      if (mon_perm.find('p') != string::npos)
+	mon_perm.replace(mon_perm.find('p'), 1, "");
+      if (mon_perm.find('s') != string::npos)
+	mon_perm.replace(mon_perm.find('s'), 1, "");
+      mon_cap_string = "allow " + mon_perm;
+
       mds_cap_string += mds_cap_string.empty() ? "" : ", ";
       mds_cap_string += "allow " + cap;
-      if (path != "/") {
-	mds_cap_string += " path=" + path;
+
+      if (filesystem != "*" && filesystem != "all") {
+	auto fs = mon->mdsmon()->get_fsmap().get_filesystem(filesystem);
+	if (!fs) {
+	  ss << "filesystem " << filesystem << " does not exist.";
+	  err = -EINVAL;
+	  goto done;
+	} else {
+	  mds_cap_string += " fsname=" + std::string(fs->mds_map.get_fs_name());
+	  mon_cap_string += " fsname=" + std::string(fs->mds_map.get_fs_name());
+	}
       }
-    }
 
-    if (filesystem != "*" && filesystem != "all") {
-      auto fs = mon->mdsmon()->get_fsmap().get_filesystem(filesystem);
-      if (!fs) {
-	ss << "filesystem " << filesystem << " does not exist.";
-	err = -EINVAL;
-	goto done;
+      if (path != "/") {
```

Comment:
shouldn't this be outside the `if (filesystem != "*" && filesystem != "all") ` statement?